File: go/010/main.go
/*
 * ProjectEuler 010
 * Go / main.go
 * copyright (c) 2015 Susisu
 *
 * Go 1.4.1
 * `go build -o bin/main main.go`
 */


package main

import "fmt"

func main() {
    n := 2000000
    
    table := make([]bool, n)
    for i := 0; i < n; i++ {
        table[i] = true
    }
    table[0] = false
    table[1] = false

    sum := 0
    for i := 2; i < n; i++ {
        if table[i] {
            for j := i * i; j < n; j += i {
                table[j] = false
            }
            sum += i
        }
    }

    fmt.Printf("%d\n", sum)
}

File: rust/010/main.rs
/*
    ProjectEuler 010
    Rust / main.rs
    copyright (c) 2015 Susisu

    Rust 1.3.0
    `rustc main.rs -o bin/main`
*/

fn main() {
    const N: usize = 2000000;
    let mut table: Vec<bool> = Vec::new();
    table.reserve(N);
    while table.len() < N {
        table.push(true);
    }
    table[0] = false;
    table[1] = false;
    let r = (N as f32).sqrt().ceil() as usize;
    for i in 2..r {
        if table[i] {
            let mut j = i * i;
            while j < N {
                table[j] = false;
                j += i;
            }
        }
    }
    let mut sum: u64 = 0;
    for i in 0..N {
        if table[i] {
            sum += i as u64;
        }
    }
    println!("{}", sum);
}
